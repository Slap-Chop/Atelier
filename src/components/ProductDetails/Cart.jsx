import React from 'react';
import Select from 'react-select';
import axios from 'axios';
import config from '../../../config.js';

class Cart extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      currentSize: 'Select Size',
      stock: {},
      inStock: false,
      stockMessage: 'Out of Stock',
      quantity: 0,
      skuId: 0,
      skuObj: {},
      notSelected: false,
    }
    this.sizeRef = React.createRef();
  }

  componentDidUpdate() {
    // console.log(this.state.stock)
    if(this.props.currentStyle.skus !== this.state.stock) {
      this.setState({stock: this.props.currentStyle?.skus, currentSize: 'Select Size'})
      if (this.props.currentStyle?.skus?.length !== 0) {
        this.setState({stockMessage: 'Select Size', inStock: true});
      }
      if (this.props.currentStyle?.skus?.null) {
        this.setState({stockMessage: 'Out of Stock', inStock: false, currentSize: 'Select Size'})
      }
    }
  }

  handleSizeChange(event) {
    let skuId = event?.value
    let skuObj = this.state.stock[skuId] || {quantity: 0, size: 'Select Size'}
    this.setState({notSelected: false});
    // setting state to have the select sku information
    if (skuId === 'N/A') {
      this.setState({sku: 0,
      skuObj: {},
      currentSize: 'Select Size',
      quantity: 0,
      })
    } else {
      this.setState({sku: skuId,
        skuObj: skuObj,
        currentSize: skuObj?.size,
        quantity: this.state.quantity > skuObj.quantity ? 1 : this.state.quantity ? this.state.quantity : 1
      })
    }
    // console.log(skuId, this.state.stock[skuId], skuObj?.size)
  }

  handleQuantChange(event) {
    // console.log(event?.target?.value)
    this.setState({quantity: event?.target?.value})
  }

  handleAddCart() {
    // console.log(this.props.currentStyle.name, this.state.quantity, this.state.currentSize)
    let productObj = {
      sku_id: this.state.sku,
      count: this.state.quantity
    };
    // console.log(productObj)
    if (this.state.currentSize === 'Select Size') {
      this.setState({notSelected: true})
      this.sizeRef.current?.focus();
    } else {
      axios.defaults.headers.common['Authorization'] = config.TOKEN;
      axios.post('https://app-hrsei-api.herokuapp.com/api/fec2/hr-rfp/cart/', productObj).then((response) => {
        console.log(`product ${this.props.currentStyle.name} added to cart with size ${this.state.currentSize}, quantity ${this.state.quantity}, sku ${this.state.sku} and response status ${response.status}`)
      })
    }
    // console.log('skus', this.props.currentStyle?.skus)

  }

  render() {
    let quantArray = [];
    let maxquant = this.state.skuObj?.quantity > 15 ? 15 : this.state.skuObj?.quantity
    for (let i = 1; i <= maxquant; i++) {
      quantArray.push(i)
    }

    let options = [{value: 'N/A', label: this.state.stockMessage}]

    if (this.state.inStock && this.state.stock) {
      let newoptions = [];
      newoptions.push({value: 'N/A', label: 'Select Size'})
      Object.entries(this.state.stock).forEach((sku, index) => {
        newoptions.push({value: sku[0], label: sku[1].size})
      })
      options = newoptions
    }

    return(
      <div className='productBorder'>
        {this.props.currentProduct?.name} > {this.props.currentStyle?.name}
      <div className='cartContainer'>
      <div className='sizeContainer'>
      {/* Size dropdown menu */}
      {this.state.notSelected && 'Please select a size'}
      {this.state.inStock &&
        <Select onChange={this.handleSizeChange.bind(this)}
      openMenuOnFocus={true}
      defaultValue={{label: 'Select Size', value: 'N/A'}}
      options={options}
      ref={this.sizeRef}
      className='sizeMenu'
      isDisabled={!this.state.inStock}/>}
      {/* check if item is in stock*/}
      {!this.state.inStock && <Select onChange={this.handleSizeChange.bind(this)}
      openMenuOnFocus={true}
      defaultValue={{label: 'Out of Stock', value: 'N/A'}}
      options={options}
      ref={this.sizeRef}
      className='sizeMenu'
      isDisabled={!this.state.inStock}/>}
        {/* <option value={'N/A'}>{this.state.stockMessage}</option> */}
      {/* map and add sizes to the dropdown */}
        {/* {this.state.stock && Object.entries(this.state.stock)?.map((sku, index) => {
        return (<option value={sku[0]} key={index}>{sku[1].size}</option>)
      })} */}


      </div>
      <div>
        {/* Quantity dropdown menu */}
        <select className='quantityMenu'
        onChange={this.handleQuantChange.bind(this)}
        disabled={this.state.currentSize === 'Select Size'}>
        {this.state.currentSize === 'Select Size' && <option>-</option>}
        {this.state.currentSize !== 'Select Size' && quantArray.map((number, index) => {
          return <option key={index} value={number}>{number}</option>
        })}
        </select>
      </div>

      </div>

      {/* Add to Cart Button */}
      {this.state.inStock && <button onClick={this.handleAddCart.bind(this)}>Add to Cart</button>}
    </div>
    )
  }


}

export default Cart;