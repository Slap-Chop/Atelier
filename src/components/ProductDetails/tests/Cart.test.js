import React from 'react';
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import Cart from '../Cart.jsx';


const testStyle = {
  'default?': true,
  name: "Forest Green & Black",
  original_price: "140.00",
  photos: [{
    thumbnail_url: "https://images.unsplash.com/photo-1501088430049-71c79fa3283e?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80",
    url: "https://images.unsplash.com/photo-1501088430049-71c79fa3283e?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=668&q=80"
  }],
  sale_price: null,
  skus: {1394769: {quantity: 8, size: 'XS'}},
  style_id: 240500
}

const diffTestStyle = {
  'default?': true,
  name: "Different Style",
  original_price: "1337.00",
  photos: [{
    thumbnail_url: "https://images.unsplash.com/photo-1501088430049-71c79fa3283e?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80",
    url: "https://images.unsplash.com/photo-1501088430049-71c79fa3283e?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=668&q=80"
  }],
  sale_price: "111.00",
  skus: {
    1394769: {quantity: 8, size: 'XS'},
    1394770: {quantity: 100, size: 'M'},
    1394771: {quantity: 19, size: 'XL'}
  },
  style_id: 240500
}

const emptyTestStyle ={
  skus: {}
}

const testProduct ={
  category: "Jackets",
  default_price: "140.00",
  description: "The So Fatigues will wake you up and fit you in. This high energy camo will have you blending in to even the wildest surroundings.",
  id: 40344,
  name: "Camo Onesie",
  slogan: "Blend in to your crowd"
}

const testCart = shallow(<Cart currentStyle={testStyle} currentProduct={testProduct}/>)

//testing render to stay constant
test('Cart should render properly', () => {
  expect(toJson(testCart)).toMatchSnapshot();
})


//testing component did update
test('Cart should update properly', () => {
  const spy = jest.spyOn(Cart.prototype, "componentDidUpdate")
  const wrapper = shallow(<Cart currentStyle={testStyle} currentProduct={testProduct}/>)
  expect(spy).not.toHaveBeenCalled();
  wrapper.setProps({currentStyle:{diffTestStyle}, currentProduct: {testProduct}})
  expect(spy).toHaveBeenCalled();
})

//testing add to Cart hidden with no stock

test('Add to cart button should be hidden if there is no stock', () => {
  const wrapper = shallow(<Cart currentStyle={emptyTestStyle} currentProduct={testProduct}/>)
  expect(wrapper.find('button')).toHaveLength(0);
  expect(wrapper.instance().state.inStock).toEqual(false);
})

test('Changing size from dropdown should call handleSizeChange', () => {
  const spy = jest.spyOn(Cart.prototype, "handleSizeChange");
  const wrapper = shallow(<Cart currentStyle={diffTestStyle} currentProduct={testProduct}/>)
  expect(spy).not.toHaveBeenCalled();
  expect(wrapper.find('select.sizeMenu')).toHaveLength(1);
  wrapper.instance().state.stock = {1394771: {quantity: 100, size: 'M'}}
  wrapper.find('select.sizeMenu').at(0).simulate('change', {
    target: {value: 1394771}
  })
  expect(spy).toHaveBeenCalled();
})

test('Changing quant from dropdown should call handleQuantChange', () => {
  const spy = jest.spyOn(Cart.prototype, "handleQuantChange");
  const wrapper = shallow(<Cart currentStyle={diffTestStyle} currentProduct={testProduct}/>)
  expect(spy).not.toHaveBeenCalled();
  expect(wrapper.find('select.quantityMenu')).toHaveLength(1);
  wrapper.instance().state.stock = {1394771: {quantity: 100, size: 'M'}}
  wrapper.find('select.quantityMenu').at(0).simulate('change', {
    target: {value: 5}
  })
  expect(spy).toHaveBeenCalled();
  expect(wrapper.instance().state.quantity).toEqual(5)
})

test('Clicking Add Cart should call handleAddCart', () => {
  const spy = jest.spyOn(Cart.prototype, "handleAddCart");
  const wrapper = shallow(<Cart currentStyle={diffTestStyle} currentProduct={testProduct}/>)
  expect(spy).not.toHaveBeenCalled();
  wrapper.instance().state.stock = {1394771: {quantity: 100, size: 'M'}}
  wrapper.find('select.sizeMenu').at(0).simulate('change', {
    target: {value: 1394771}
  })
  wrapper.find('button').simulate('click');
  expect(spy).toHaveBeenCalled();
})