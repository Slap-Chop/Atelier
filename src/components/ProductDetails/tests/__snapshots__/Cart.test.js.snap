// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart should render properly 1`] = `
<div
  className="productBorder"
>
  Camo Onesie
   &gt; 
  Forest Green & Black
  <div
    className="cartContainer"
  >
    <div
      className="sizeContainer"
    >
      <ForwardRef
        className="sizeMenu"
        defaultValue={
          Object {
            "label": "Select Size",
            "value": "N/A",
          }
        }
        isDisabled={true}
        onChange={[Function]}
        openMenuOnFocus={true}
        options={
          Array [
            Object {
              "label": "Out of Stock",
              "value": "N/A",
            },
          ]
        }
      />
    </div>
    <div>
      <select
        className="quantityMenu"
        disabled={true}
        onChange={[Function]}
      >
        <option>
          -
        </option>
      </select>
    </div>
  </div>
</div>
`;
